akka-http/src/main/scala/akka/http/javadsl/server/Directives.scala

```diff
@@ -26,4 +28,96 @@ object Directives extends AllDirectives {
 
   @varargs override def getFromBrowseableDirectories(directories: String*): Route =
     super.getFromBrowseableDirectories(directories: _*)
+
+  /**
+   * Composes two 0-argument directives which share the same inner Route. This is equivalent to
+   * `first(inner).orElse(second(inner))`
+   * Usage example:
+   * `anyOf(this::get, this::post, () -> complete("hi")))`
+   * @param first the first 0-argument directive
+   * @param second the second 0-argument directive
+   * @param inner the inner route Producer
+   * @return the resulting route
+   */
+  // @akka.annotation.ApiMayChange // FIXME use the real ones once Akka dependency bumped
+  def anyOf(first: Function[Supplier[Route], Route], second: Function[Supplier[Route], Route], inner: Supplier[Route]): Route = {
+    first.apply(inner).orElse(second.apply(inner))
+  }
+
+  /**
+   * Composes two 1-argument directives (directives whose inner route takes a parameter) which share the same inner route.
+   * This is the same as `first(inner).orElse(second(inner))`.
+   * This can be used also for directives that take an additional parameter if the directive is partially applied, ie. parameter
+   * Usage example:
+   * `anyOf(bindParameter(this::path, "bar"), bindParameter(this::path, "baz"), () -> complete("bar or baz"))`
+   * @param first the first 1-argument directive
+   * @param second the second 1-argument directive
+   * @param inner the inner route that takes one argument
+   * @tparam A the type of the parameter the directives extract and the inner route takes
+   * @return the resulting route
+   */
+  // @akka.annotation.ApiMayChange // FIXME use the real ones once Akka dependency bumped
+  def anyOf[A](first: Function[Function[A, Route], Route], second: Function[Function[A, Route], Route], inner: Function[A, Route]): Route = {
+    first.apply(inner).orElse(second.apply(inner))
+  }
+
+  /**
+   * Nests two 0-argument directives together
+   * Usage example:
+   * `allOf(bindParameter(this::pathPrefix, "alice"), bindParameter(this::path, "bob"), () -> complete("Charlie!"))`
+   * @param first the 0-argument directive (outer one)
+   * @param second the 0-argument directive (inner one)
+   * @param inner the inner route function
+   * @return the resulting route
+   */
+  // @akka.annotation.ApiMayChange // FIXME use the real ones once Akka dependency bumped
+  def allOf(first: Function[Supplier[Route], Route], second: Function[Supplier[Route], Route], inner: Supplier[Route]): Route = {
```

Comment:
Yeah we can add more later. Would use the sbt-boilerplate plugin then - note that we do this in many places already.

I agree it's likely enough for now